appscale/tools/agents/azure_agent.py

```diff
@@ -700,81 +700,148 @@ def terminate_instances(self, parameters):
+    instances_to_delete = parameters[self.PARAM_INSTANCE_IDS]
     AppScaleLogger.verbose("Terminating the vm instance(s) '{}'".
-                           format(instance_ids), verbose)
-    compute_client = ComputeManagementClient(credentials, subscription_id)
+                           format(instances_to_delete), verbose)
 
-    # Delete the virtual machine scale sets created.
+    compute_client = ComputeManagementClient(credentials, subscription_id)
     vmss_list = compute_client.virtual_machine_scale_sets.list(resource_group)
+    downscale = parameters['autoscale_agent']
+
+    # On downscaling of instances, we need to delete the specific instance
+    # from the Scale Set.
+    if downscale in ['True', True]:
+      # Delete the scale set virtual machines matching the given instance ids.
+      vmss_vm_delete_threads = []
+      for vmss in vmss_list:
+        vm_list = compute_client.virtual_machine_scale_set_vms.list(
+          resource_group, vmss.name)
+        for vm in vm_list:
+          if vm.name in instances_to_delete:
+            instances_to_delete.remove(vm.name)
+            thread = threading.Thread(target=self.delete_vmss_instance,
+                                      args=(compute_client, parameters,
+                                            vmss.name, vm.instance_id))
+            thread.start()
+            vmss_vm_delete_threads.append(thread)
+
+      for delete_thread in vmss_vm_delete_threads:
+        delete_thread.join()
+
+      AppScaleLogger.log("Virtual machine(s) have been successfully downscaled.")
+      AppScaleLogger.log("Cleaning up any Scale Sets, if needed ...")
+      vmss_delete_threads = []
+      for vmss in vmss_list:
+        vm_list = compute_client.virtual_machine_scale_set_vms.list(
+          resource_group, vmss.name)
+        vm_names = []
+        for vm in vm_list:
+          vm_names.append(vm.name)
+
+        if not vm_names:
+          thread = threading.Thread(
+            target=self.delete_virtual_machine_scale_set, args=(
+              compute_client, parameters, vmss.name))
+          thread.start()
+          vmss_delete_threads.append(thread)
+
+      for delete_thread in vmss_delete_threads:
+        delete_thread.join()
+      return
+
+    # On appscale down --terminate, we delete all the Scale Sets within the
+    # resource group specified, as it is faster than deleting the individual
+    # instances within each Scale Set.
+    delete_ss_instances = []
     vmss_delete_threads = []
-    ss_instance_ids_to_delete = []
     for vmss in vmss_list:
       vm_list = compute_client.virtual_machine_scale_set_vms.list(
         resource_group, vmss.name)
       for vm in vm_list:
-        ss_instance_ids_to_delete.append(vm.name)
-      thread = threading.Thread(
-        target=self.delete_virtual_machine_scale_set, args=(
-          compute_client, resource_group, verbose, vmss.name))
+        delete_ss_instances.append(vm.name)
```

Comment:
The list of instances to delete consists of both the scale set vms and the regular (load balancer) vms. I need this list of scale set vms so that I only give the remaining vms to delete to the delete_virtual_machine method. In this line over here: https://github.com/AppScale/appscale-tools/pull/621/files#diff-6b911255ca065691da775e67ef91b1e2R768